Source/projections/Builder/ProjectionClassBuilder.ts

```diff
@@ -0,0 +1,161 @@
+            const [hasEventType, eventType] = this.tryGetEventTypeFromMethod(method, eventTypes);
+
+            if (!hasEventType) {
+                allMethodsValid = false;
+                logger.warn(`Could not create projection method ${method.name} in projection ${this._projectionType.name} because it is not associated to an event type`);
+                continue;
+            }
+
+            const onMethod = this.createProjectionMethod(method, container);
+            const keySelector = method.keySelector;
+
+            if (events.has(eventType!)) {
+                allMethodsValid = false;
+                logger.warn(`Event handler ${this._projectionType.name} has multiple projection methods handling event type ${eventType}`);
+                continue;
+            }
+            events.set(eventType!, [onMethod, keySelector]);
+        }
+        return allMethodsValid;
+    }
+
+    private createProjectionMethod(method: OnDecoratedMethod, container: IContainer): ProjectionCallback<any> {
+        return (instance, event, projectionContext) => {
+            const result = method.method.call(instance, event, projectionContext);
+            if (result instanceof DeleteReadModelInstance) {
+                return result;
+            }
+            return instance;
+        };
+    }
+
+    private tryGetEventTypeFromMethod(method: OnDecoratedMethod, eventTypes: IEventTypes): [boolean, EventType | undefined] {
+        if (this.eventTypeIsId(method.eventTypeOrId)) {
+            return [
+                true,
+                new EventType(
+                    EventTypeId.from(method.eventTypeOrId),
+                    method.generation ? Generation.from(method.generation) : Generation.first)
+            ];
+        } else if (!eventTypes.hasFor(method.eventTypeOrId)) {
+            return [false, undefined];
+        } else {
+            return [true, eventTypes.getFor(method.eventTypeOrId)];
+        }
+    }
+
+    private eventTypeIsId(eventTypeOrId: Constructor<any> | EventTypeId | Guid | string): eventTypeOrId is EventTypeId | Guid | string {
+        return eventTypeOrId instanceof EventTypeId || eventTypeOrId instanceof Guid || typeof eventTypeOrId === 'string';
+    }
+
+    private ThrowIfReadModelAlreadyRegistered(types: ProjectionDecoratedType[]): void {
+        const readModels = types
+            .map(type => type.readModel);
+        // make it into a Set so that we only get the unique duplicate values
+        const duplicateReadmodels = new Set(
+            readModels.filter((readModel, index) => readModels.indexOf(readModel) !== index));
+
+        if (duplicateReadmodels) {
+            for (const duplicateReadModel of new Set(duplicateReadmodels)) {
+                const duplicateTypes = types.filter(type => type.readModel === duplicateReadModel);
+                throw new ReadModelAlreadyRegistered(duplicateReadModel, duplicateTypes);
+            }
+        }
```

Comment:
This code could be simplified a bit to not need a comment, and the set is made twice which seems excessive.

Reply:
Removed the extra `Set` and unnecessary `if`, as `!!new Set() === true`

Reply:
Actually just removing this